### ISSUE TITLE
[tcgc] add client accessor support to `client.tsp`

### ISSUE THREAD

**Issue by iscai-msft** [Azure/typespec-azure#1715]
Currently, TCGC doesn't support public client accessors from parent to subclient. 

The canonical use case is something like `storage`, where there is the access from parent `BlobServiceClient` to child `ContainerClient`, with `BlobServiceClient.getContainerClient(containerName: string)` as a client accessor between the two.

By default, with parents and subclients, unless an explicit client accessor method is defined, there is no generated public accessor method from parent to subclient. If a user wants a public accessor method from parent to subclient, they will define it in `client.tsp` like so

```tsp
// client.tsp

op getContainerClient(containerName: string): Azure.Storage.Blob.Container;
```

If you want to define a client accessor and have `access` still be `internal`, the `@access` decorator can still be applied. This use case would be if there are special internal parameters that you have to pass to the subclient that we don't want users to deal with

## TODO

1. We need to make sure that clients / namespaces / interfaces can be listed as a return type in an operation. This might require working with the typespec team.

**Comment by iscai-msft** [comment]
I still feel that having an explicit definition is better, instead of going down a logic map way of initialization and accessor. I guess the root of the issue is, besides storage and some other individual examples, we don't have a good feel if a public client has a public subclient, whether we have to add a separate ctor.

The thing with explicitly defining is we don't have to tie ourselves to logic. I don't think we're 100% blocked by typespec support actually, we can return namespaces and interfaces from operations (ex [here](https://typespec.io/playground/?c=CkBzZXJ2aWNlCm5hbWVzcGFjZSBBIHsKCiAgb3Agc3ViTsgXQ2xpZW50KCk6IEI7yR5JbnRlcmbNHkM7xD3KTkIgewogIH3EFWnILiBDxxV9Cg%3D%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)). Finally, I also think we want to give support to spec authors to explicitly state what parameters should get passed from parent to child etc

**Comment by weidongxu-microsoft** [comment]
First, wondering whether we have different logic for accessor without parameter and accessor with parameter. My following comment is on assumption that we didn't distinguish the 2. Both are just client accessor.

Now, wondering for how many languages, this statement is currently true?
> By default, with parents and subclients, unless an explicit client accessor method is defined, there is no generated public accessor method from parent to subclient.

In this default mode, then how the subclient get initialized?

Java actually fit this (as default flatten-client mode), but I think Java is really an exception -- in such mode the subclient is initialized via a Builder of parent client.

I think for .NET, there is always a `.getSubClient()` from parent client to subclient (at least for cases of no additional parameters required in subclient). @pshao25 Is it correct?

I do not have good recommendation.
- If we enforce that service must write `op getContainerClient(containerName: string);` in client.tsp, it seems redundant in most cases, and won't easily work in unbranded (the default being all subclient not initializable and not usable?)
- On the other hand, it may not be easy to devise another "default" mode. For `@cilentInitialization`, without this decorator there is still some default (is it correct?) for initializing the client. If we had similar "default" in client accessor, it may not be apparent to dev who writes the typespec.

PS: currently Java (when in subclient mode) would not require dev write additional client accessor method. The client accessor method is automatically generated (with a name) parameters in the `@cilentInitialization` -- logic be: if `containerName` param is required in `@cilentInitialization`, it is also a required param in client accessor method.

**Comment by tadelesh** [comment]
> * If we enforce that service must write `op getContainerClient(containerName: string);` in client.tsp, it seems redundant in most cases, and won't easily work in unbranded (the default being all subclient not initializable and not usable?)

agree. i think in most cases, the explicit definition is redundant. actually, i got some feedback that the way we build our client hierarchy is not so convenient for emitter, as they only need the hierarchy info, no need to know the method info. since different language do have different way to deal with the client ctor or client accessor, i prefer just to construct the hierarchy with parent or children linkage, instead of to redefine the accessor method. with `@clientInitialization`, user could bump the method parameter to client and make the ctor public. with maybe `@clientAccessor`, user could redefine the client accessor method and make it public.

**Comment by iscai-msft** [comment]
i think this is a good topic for today's sync, let me make note of it
